internal/statesync/dispatcher.go

```diff
@@ -281,6 +285,25 @@ func (l *peerList) Append(peer types.NodeID) {
 	}
 }
 
+func (l *peerList) Extend(newPeers []types.NodeID) {
+	l.mtx.Lock()
+	defer l.mtx.Unlock()
+
+	set := make(map[types.NodeID]struct{}, len(l.peers))
+
+	for _, p := range l.peers {
+		set[p] = struct{}{}
+	}
+
+	for _, p := range newPeers {
+		if _, ok := set[p]; ok {
+			continue
+		}
+
+		l.unsafeAppend(p)
```

Comment:
Can we be processing a peer update while this happens?

Reply:
That seems unlikely but there's a mutex that would protect this structure in both cases, so I think it's safe even if it does happen.